commit 334fe134447d37cca9f3193351871e2aea635ac9
Author: Michal Hocko <mhocko@suse.com>
Date:   Wed Feb 22 15:46:01 2017 -0800

    mm, vmscan: consider eligible zones in get_scan_count
    
    commit 71ab6cfe88dcf9f6e6a65eb85cf2bda20a257682 upstream.
    
    get_scan_count() considers the whole node LRU size when
    
     - doing SCAN_FILE due to many page cache inactive pages
     - calculating the number of pages to scan
    
    In both cases this might lead to unexpected behavior especially on 32b
    systems where we can expect lowmem memory pressure very often.
    
    A large highmem zone can easily distort SCAN_FILE heuristic because
    there might be only few file pages from the eligible zones on the node
    lru and we would still enforce file lru scanning which can lead to
    trashing while we could still scan anonymous pages.
    
    The later use of lruvec_lru_size can be problematic as well.  Especially
    when there are not many pages from the eligible zones.  We would have to
    skip over many pages to find anything to reclaim but shrink_node_memcg
    would only reduce the remaining number to scan by SWAP_CLUSTER_MAX at
    maximum.  Therefore we can end up going over a large LRU many times
    without actually having chance to reclaim much if anything at all.  The
    closer we are out of memory on lowmem zone the worse the problem will
    be.
    
    Fix this by filtering out all the ineligible zones when calculating the
    lru size for both paths and consider only sc->reclaim_idx zones.
    
    The patch would need to be tweaked a bit to apply to 4.10 and older but
    I will do that as soon as it hits the Linus tree in the next merge
    window.
    
    Link: http://lkml.kernel.org/r/20170117103702.28542-3-mhocko@kernel.org
    Fixes: b2e18757f2c9 ("mm, vmscan: begin reclaiming pages on a per-node basis")
    Signed-off-by: Michal Hocko <mhocko@suse.com>
    Tested-by: Trevor Cordes <trevor@tecnopolis.ca>
    Acked-by: Minchan Kim <minchan@kernel.org>
    Acked-by: Hillf Danton <hillf.zj@alibaba-inc.com>
    Acked-by: Mel Gorman <mgorman@suse.de>
    Acked-by: Johannes Weiner <hannes@cmpxchg.org>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Index: linuxtest/dummy/rpi_1256_334fe134447d37cca9f3193351871e2aea635ac9.txt
===================================================================
--- /dev/null
+++ linuxtest/dummy/rpi_1256_334fe134447d37cca9f3193351871e2aea635ac9.txt
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
